src/main/kotlin/com/sourcegraph/cody/edit/LensesService.kt

```diff
@@ -0,0 +1,162 @@
+package com.sourcegraph.cody.edit
+
+import com.intellij.openapi.application.runInEdt
+import com.intellij.openapi.components.Service
+import com.intellij.openapi.components.service
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.fileEditor.FileEditorManager
+import com.intellij.openapi.fileEditor.OpenFileDescriptor
+import com.intellij.openapi.project.Project
+import com.sourcegraph.cody.Icons
+import com.sourcegraph.cody.agent.protocol_generated.ProtocolCodeLens
+import com.sourcegraph.cody.agent.protocol_generated.Range
+import com.sourcegraph.cody.edit.widget.LensAction
+import com.sourcegraph.cody.edit.widget.LensHotkey
+import com.sourcegraph.cody.edit.widget.LensIcon
+import com.sourcegraph.cody.edit.widget.LensLabel
+import com.sourcegraph.cody.edit.widget.LensSpinner
+import com.sourcegraph.cody.edit.widget.LensWidgetGroup
+import com.sourcegraph.utils.CodyEditorUtil
+import javax.swing.Icon
+
+interface LensListener {
+  fun onLensesUpdate(
+      uri: String,
+      taskId: String?,
+      lensWidgetGroup: LensWidgetGroup?,
+      codeLenses: List<ProtocolCodeLens>
+  )
+}
+
+@Service(Service.Level.PROJECT)
+class LensesService(val project: Project) {
+  private var lensGroups = mutableMapOf<String, Map<String, Pair<Range, LensWidgetGroup>>>()
+
+  private val listeners = mutableListOf<LensListener>()
+
+  fun addListener(listener: LensListener) {
+    listeners.add(listener)
+  }
+
+  fun removeListener(listener: LensListener) {
+    listeners.remove(listener)
+  }
+
+  fun updateLenses(uri: String, codeLens: List<ProtocolCodeLens>) {
+    runInEdt {
+      val vf = CodyEditorUtil.findFileOrScratch(project, uri) ?: return@runInEdt
+      val fileDesc = OpenFileDescriptor(project, vf)
+      val editor =
+          FileEditorManager.getInstance(project).openTextEditor(fileDesc, true) ?: return@runInEdt
+
+      val newLensGroups =
+          codeLens
+              .groupBy { it.range }
+              .map { (range, codeLensesModels) ->
+                val taskId = codeLensesModels.firstNotNullOf { getTaskId(it) }
+                taskId to (range to createLensGroup(editor, codeLensesModels))
+              }
+              .toMap()
+
+      lensGroups[uri]?.values?.forEach { it.second.dispose() }
```

Comment:
The Jetbrains documentation [says](https://plugins.jetbrains.com/docs/intellij/disposers.html#diagnosing-disposer-leaks) that Disposable.dispose() should never be called directly